src/Examine.AzureDirectory/AzureLuceneDirectory.cs

```diff
@@ -40,58 +44,107 @@ public class AzureDirectory : Lucene.Net.Store.Directory
+            IsReadOnly = isReadOnly;
+            _storageAccountConnectionString = connectionString;
             CacheDirectory = cacheDirectory;
             _containerName = containerName.ToLower();
+            _lockFactory = GetLockFactory();
+            RootFolder = NormalizeContainerRootFolder(rootFolder);
 
-            _lockFactory = isReadOnly
-                ? (LockFactory)new NoopLockFactory()
-                : new MultiIndexLockFactory(new AzureDirectorySimpleLockFactory(this), CacheDirectory.LockFactory);
+            EnsureContainer();
+            _azureIndexOutputFactory = GetAzureIndexOutputFactory();
+            _azureIndexInputFactory = GetAzureIndexInputFactory();
+            GuardCacheDirectory(CacheDirectory);
+            CompressBlobs = compressBlobs;
+        }
+        protected virtual IAzureIndexInputFactory GetAzureIndexInputFactory()
+        {
+            return new AzureIndexInputFactory();
+        }
+        protected virtual IAzureIndexOutputFactory GetAzureIndexOutputFactory()
+        {
+            return new AzureIndexOutputFactory();
+        }
+
+        protected virtual void GuardCacheDirectory(Lucene.Net.Store.Directory cacheDirectory)
+        {
+            if (cacheDirectory == null) throw new ArgumentNullException(nameof(cacheDirectory));
+        }
 
+        protected string NormalizeContainerRootFolder(string rootFolder)
+        {
             if (string.IsNullOrEmpty(rootFolder))
-                RootFolder = string.Empty;
-            else
-            {
-                rootFolder = rootFolder.Trim('/');
-                RootFolder = rootFolder + "/";
-            }
+                return string.Empty;
+            rootFolder = rootFolder.Trim('/');
+            rootFolder = rootFolder + "/";
+            return rootFolder;
+        }
 
-            _blobClient = storageAccount.CreateCloudBlobClient();
-            EnsureContainer();
-            CompressBlobs = compressBlobs;
+        protected virtual LockFactory GetLockFactory()
+        {
+            return IsReadOnly ? (LockFactory)new NoopLockFactory()
+                : new MultiIndexLockFactory(new AzureDirectorySimpleLockFactory(this), CacheDirectory.LockFactory);
+        }
+        protected virtual BlobClient GetBlobClient(string blobName)
+        {
+            return new BlobClient(_storageAccountConnectionString, _containerName, blobName);
+        }
+        protected virtual BlobContainerClient GetBlobContainerClient(string containerName)
+        {
+            return new BlobContainerClient(_storageAccountConnectionString, _containerName);
         }
 
         public string RootFolder { get; }
-        public CloudBlobContainer BlobContainer => _blobContainer;
         public bool CompressBlobs { get; }
-        public Lucene.Net.Store.Directory CacheDirectory { get; }
 
+        public Lucene.Net.Store.Directory CacheDirectory { get; protected set; }
```

Comment:
@Shazwazza it can't be readonly as it is used an swapped by reaonly, it is why it is protected. if you check there logic: 
https://github.com/Shazwazza/Examine/blob/c4f94832ea385ff559c0f6626653f9a7bb0714cb/src/Examine.AzureDirectory/AzureReadOnlyLuceneDirectory.cs#L57
To avoid corruption of indexes :)